docs/proposals/cli-improvements.md

```diff
@@ -0,0 +1,129 @@
+
+The behaviour of Java protoc CLI is obvious since it generates classes for all the proto files inside the pre-allocated proto directory. The behaviour of Go protoc CLI is as follows.
+
+```sh
+protoc --go_out=go2 --go-grpc_out=go2 proto/**.proto --proto_path=proto
+```
+In Golang, we can input proto files as a wildcard, and then it generates all the necessary stubs [1].
+
+#### Proposed Solution
+There are two ways we can address this problem, one way to follow the Java approach and the other way is to follow the Go approach.
+##### Solution 01
+We keep the current CLI interface as it is. When someone inputs a proto directory as an input, we iterate through the nested directories and generate all the necessary stub files.
+```sh
+bal grpc –input proto/ –output ballerina-out
+```
+##### Solution 02
+Allow users to use wild cards as Go did.
+```sh
+bal grpc –input proto/**.proto –output ballerina-out
+```
+
+### Problem 02 - Centralized Proto Descriptor
+#### Problem Definition
+In our current Ballerina gRPC library, the generated stub contains a large single descriptor (root descriptor + descriptor map), which is related to the service and the messages that the service uses. But, when we address problem 01 and create several stubs for each proto file in a directory, maintaining such a huge descriptor will be a difficult task. Therefore, we should decentralize our descriptor (especially the descriptor map) like other languages. For example, Java has the descriptor details in the message class itself and Golang has it in the generated struct.
+E.g.,
+```go
+…
+var file_messages2_proto_rawDesc = []byte{
+	0x0a, 0x0f, 0x6d, 0x65, 0x73, 0x73, 0x61, 0x67, 0x65, 0x73, 0x32, 0x2e, 0x70, 0x72, 0x6f, 0x74,
+	0x6f, 0x22, 0x1c, 0x0a, 0x08, 0x4d, 0x65, 0x73, 0x73, 0x61, 0x67, 0x65, 0x32, 0x12, 0x10, 0x0a,
+	0x03, 0x6d, 0x73, 0x67, 0x18, 0x01, 0x20, 0x01, 0x28, 0x09, 0x52, 0x03, 0x6d, 0x73, 0x67, 0x42,
+	0x17, 0x5a, 0x15, 0x65, 0x78, 0x61, 0x6d, 0x70, 0x6c, 0x65, 0x2e, 0x63, 0x6f, 0x6d, 0x2f, 0x6d,
+	0x65, 0x73, 0x73, 0x61, 0x67, 0x65, 0x73, 0x32, 0x62, 0x06, 0x70, 0x72, 0x6f, 0x74, 0x6f, 0x33,
+}
+…
+```
+#### Proposed Solution
+The proposed solution is to introduce a message-level annotation for the generated records. The descriptor value is related to a particular Proto file. For example, for each proto file (i.e. stub) there will be one descriptor variable. Then, all the recorded in the stub file use the same descriptor value.
+
+```proto
+message Person {
+  string name = 1;
+  int32 id = 2;
+  string email = 3;
+}
+```
+```ballerina
+const string descriptorName = "0A2736385F73696D706C655F...";
+
+@protobuf:Descriptor { value: descriptorName }
+public type Person record {|
+    string name = “”;
+    int id = 0;
+    string officeEmail = “”;
+|};
+```
+Note: Here the descriptor value should be unique. It can be generated as follows:
+```sh
+descriptorName := <PROTO_FILE_NAME_IN_CAMEL_CASE>Descriptor
```

Comment:
Since this is a compile-time constant, shall we make it uppercase?

Reply:
Sure, +1

Reply:
Updated in here, https://github.com/ballerina-platform/module-ballerina-grpc/pull/779/commits/59176096df885bcf66f8947009cfd84372395dd9